modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -210,14 +213,95 @@
         throw failure;
     }
 
+    /**
+     * Send request and handle response asynchronously.
+     */
+    public <T> IgniteClientFuture<T> serviceAsync(
+            ClientOperation op,
+            Consumer<PayloadOutputChannel> payloadWriter,
+            Function<PayloadInputChannel, T> payloadReader
+    ) throws ClientException, ClientError {
+        CompletableFuture<T> fut = new CompletableFuture<>();
+
+        ClientChannel ch = channel();
+
+        ch.serviceAsync(op, payloadWriter, payloadReader).handle((res, err) ->
+                handleServiceAsync(op, payloadWriter, payloadReader, fut, null, null, ch, res, err));
+
+        return new IgniteClientFutureImpl<>(fut);
+    }
+
+    /**
+     * Handles serviceAsync results and retries as needed.
+     */
+    private <T> Object handleServiceAsync(ClientOperation op,
+                                          Consumer<PayloadOutputChannel> payloadWriter,
+                                          Function<PayloadInputChannel, T> payloadReader,
+                                          CompletableFuture<T> fut,
+                                          ClientConnectionException failure,
+                                          AtomicInteger chIdx,
+                                          ClientChannel ch,
+                                          T res,
+                                          Throwable err) {
+        if (err == null) {
+            fut.complete(res);
+            return null;
+        }
+
+        if (err instanceof ClientConnectionException) {
```

Comment:
I think, only those operations that do not modify cache state should be retried or we can get unexpected results, if the operation was successfully performed, but we were unable to receive response.

Also, it is OK to retry if failure happens on send.

Reply:
I agree, but this behavior is consistent with sync API, I don't think we should have different behavior for sync and async API.

Reply:
Well, OK then